add array_scan, cumulative_sum

catoverdrive opened the pull request:
@konradjk 

fixes #4217.

ArrayScan mimics Scala behavior where it returns an array of length `n+1`, including both the base case and the last value.

tpoterba commented:
wow, that was fast.

catoverdrive commented:
turns out when you can copy most of the code from `ArrayFold`...

catoverdrive commented on python/hail/expr/functions.py:
```diff
@@ -3114,6 +3172,41 @@ def sum(collection, filter_missing: bool = True) -> NumericExpression:
+    """
+    if filter_missing:
+        a = a.filter(hl.is_defined)
+    return a.scan(hl.literal(0, a.dtype.element_type),
```
I think that seems reasonable, as long as evaluating f twice is relatively fast/painless

tpoterba replied:
it's just in Python - I'm not worried about time so much as weird errors getting generated.

tpoterba replied:
though we can check it as follows: 

```
evaluate f

if body.dtype == zero.dtype:
  proceed
else:
  coercer = coercer.from_dtype(body.dtype) # or whatever this is
  if coercer.can_coerce(zero.dtype):  # or whatever
    coerced_zero = coercer.coerce(zero)
    re evaluate body with coerced_zero and proceed
  else:
    raise the types-don't-match error
```

tpoterba replied:
ah, here's the potential problem -- what if evaluating body with `coerced_zero` produces a different type? I guess if it does that we can panic to the types-dont-match error

catoverdrive replied:
yep. I now have it trying to coerce `body` to the zero type first, since that doesn't require re-evaluation of `f`, and then trying to coerce `zero`, and checking if the new `body` is coercable to the new `zero`. Hopefully that catches common conversion problems without getting stuck in a loop trying to coerce `zero` and re-evaluate `f`, but if it doesn't work I throw an error with the original types.

catoverdrive commented on src/main/scala/is/hail/expr/ir/PruneDeadFields.scala:
```diff
@@ -670,6 +670,16 @@ object PruneDeadFields {
+      case ArrayScan(a, zero, accumName, valueName, body) =>
+        val aType = a.typ.asInstanceOf[TArray]
+        val zeroEnv = memoizeValueIR(zero, zero.typ, memo)
+        val bodyEnv = memoizeValueIR(body, body.typ, memo)
```
that makes sense. I think I could do this for ArrayFold, too, right?

tpoterba replied:
ah. 

you're totally right.

Let's punt on this for now.

catoverdrive commented:
@tpoterba I addressed all the comments except maybe the one about more aggressive pruning of the fields, where it's not super obvious to me what the thing to do is.
